Handle USB daemon changes

pentadoc opened the pull request:
The UIVM now interacts with the new USB daemon correctly. 

OXT-236

Requires commits to be pulled: 
- jean-edouard/vusb-daemon: ab6951b
- jean-edouard/idl: 893c815

pentadoc committed:
USB daemon behavior handling

UI now handles the behavior of the new usb daemon.

OXT-448

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

pentadoc committed:
Refresh VM Details usb devices

SV-606

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

pentadoc committed:
Refresh Device window when usb device plugged in

SV-606

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

pentadoc committed:
Adapt to more frequent usb changed signals

- Refresh stickyChanged in usb manipulation callbacks
- Move usbBusy singleton moved to the end of the finish callback

Addendum to SV-606

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

pentadoc committed:
Fix usb refresh timing

Embedded Midori seems to schedule async functions differently than Chrome or desktop Midori.  This patch fixes that timing issue but will most likely need to be reverted if Chromium is used in the future.

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

pentadoc committed:
Fixing CD behavior

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

pentadoc committed:
Fixing CD behavior

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

pentadoc committed:
USB interface layer refactor

This proxies the usb interface to ensure timing is correct on all USB actions.  The proxy was initially in the hostModel but was moved to it's own file.

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

pentadoc committed:
Remove unnecessary UI precaution

With the USB daemon acting predictably we no longer need to prevent the Assignment selects from being changed due to the sticky setting.

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

pentadoc committed:
Fix migration-induced error

Forgot to rename self to XUICache.Host

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

Marsland commented:
+1

rneilturner commented:
@pentadoc This is a long way forward in getting the UIVM in-sync with the new vusb daemon. As I understand it, there are still some issues you have noticed. Could you create a new OXT ticket describing some of these issues and amend this PR with this ticket? Otherwise, looks good.

jean-edouard commented on widgets/xenclient/templates/VMDetails.html:
```diff
@@ -103,14 +103,17 @@
                                 </td>
                                 <td>
-                                    <button class="usbButton" templateType="citrix.common.Button" dojoAttachEvent="onClick: onUsbDetach">${DISCONNECT}</button>
+                                    <button id="usb_button_%dev_id%"class="usbButton" templateType="citrix.common.Button" dojoAttachEvent="onClick: onUsbDetach">${DISCONNECT}</button>
```
Missing space

jean-edouard commented on dist/script/models/vmModel.js:
```diff
@@ -267,11 +268,16 @@ XenClient.UI.VMModel = function(vm_path) {
+
+                device.assigned_uuid = self.uuid;
+                interfaces.usb.set_usbDevice(device, self.refreshUsb);
+                //self.setUsbDeviceSticky(device.dev_id, device.getSticky);
```
When you comment out code, please explain why in a comment, an how it could ever be useful again, or just remove it.

pentadoc committed:
Clean up changes

Fix spacing issues, remove unneeded commented code, explain remaining comment code.

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

pentadoc committed:
Comment fix

Further describe the changes pertaining to XUICache.USB in the hostModel and vmModel.

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

pentadoc committed:
Fix Typo

pentadoc commented:
@jean-edouard I think I addressed all issues pertaining to your comments

@rneilturner After a BIOS update no serious issues remain in OXT. 

Solves issues described in OXT-448.

jean-edouard commented:
+1

rossphilipson commented:
@jean-edouard did you do a build and test these changes?

jean-edouard commented:
Yes, http://openxt.ainfosec.com:81/builds/master/ext-dev-351-master/

pentadoc committed:
Fix usb race condition

Usb tries to update before daemon has settled on Dell Venue undock.

Signed-off-by: Andrew 'Doc' Docherty <doc@coveycs.com>

pentadoc commented:
@jean-edouard @rossphilipson I will have more USB changes coming down the pike and I would feel better if it was in another pull request instead of tacked onto the end of this one.  Any idea when this can be merged?

jean-edouard commented:
Done

rneilturner commented:
FYI, this commit will not work very well without https://github.com/OpenXT/vusb-daemon/pull/9.

jctemkin commented:
This commit also requires OpenXT/idl#14.
